Man/ms-intersect.1
.TH ms-intersect 1
.SH NAME    \" Section header
.PP

ms-intersect - Report genes common to some species and missing from others

\" Convention:
\" Underline anything that is typed verbatim - commands, etc.
.SH SYNOPSIS
.PP
.nf 
.na 
ms-intersect file.gff3 file.gff3 [--diverged file.gff3 file.gff3]
.ad
.fi

.SH OPTIONS
.TP
\fB--diverged\fR
Begin list of diverged species, each of which is individually compared to
the intersection of the initial group.

\" Optional sections
.SH "DESCRIPTION"
.B ms-intersect
finds the intersection of gene groups in a list of GFF3 files such as those
output by ms-extract(1).  I.e., it produces a list of genes neighboring a
gene of interest (GOI) common to 2 or more species.  This is useful since
neighboring genes often contain regulatory elements for the GOI in their
introns.

Each GFF3 file should contain a small group of genes surrounding a gene
of interest (GOI) for a given species.

The filename of the GFF3 must be of the form

species-gene-chrom-start.gff3

where gene is the gene name (not ID), chrom is the chromosome name or
number, and start is the starting position of the gene on the chromosome.
This is the filename format produced by ms-extract(1).

Starting with the second GFF3 file, the intersection of genes in the file
will all previous gene groups is output.  Hence, the order in which files
are given on the command line determines the progression of intersection
output, though the final intersection of all species will be the same
regardless of order, since it represents a list of genes shared by all
species.

If the --diverged flag is present, each GFF3 file after it is compared
independently to the intersection of all groups before --diverged.  This
allows the user to see what genes are present in the neighboris shared by each of the diverged species

.SH EXAMPLE

The following command shows that 3 fish species all have rbm27, rad50,
and drd1b near the pou4f3 gene, while the mammals are missing rad50 and
drd1b.  The absence of the latter two genes may explain differences in
gene regulation between fish and mammals, since they could contain
regulator elements (e.g. enhancers) for pou4f3 in their introns.

.nf
.na
ms-intersect \\
    Danio_rerio-pou4f3-9-54290896.gff3 \\
    Oryzias_latipes-pou4f3-14-7315063.gff3 \\
    Takifugu_rubripes-pou4f3-15-5176582.gff3 \\
    --diverged \\
    Homo_sapiens-pou4f3-5-146338839.gff3 \\
    Mus_musculus-pou4f3-18-42527604.gff3 \\
    Rattus_norvegicus-pou4f3-18-34390205.gff3

====================
pou4f3
====================

Neighboring genes conserved among the original group:

Species              Ne Co Ch  Intersection with all previous genes
Danio_rerio           8  *  9  (NA)
Oryzias_latipes       8  3 14  rbm27 rad50 drd1b
Takifugu_rubripes     8  3 15  rbm27 rad50 drd1b

Genes conserved in each species through divergence from the group above:

Mus_musculus          8  1 18  rbm27
Rattus_norvegicus     8  1 18  rbm27
Homo_sapiens          8  1  5  rbm27
.ad
.fi

.SH "SEE ALSO"
ms-extract(1), ms-stack(1), ms-plot(1)

.SH BUGS
Please report bugs to the author and send patches in unified diff format.
(man diff for more information)

.SH AUTHOR
.nf
.na
J. Bacon


ms-intersect.c
/***************************************************************************
 *  Description:
 *      Locate major divergence in gene regions along provided
 *      inputs.
 *
 *  History: 
 *  Date        Name        Modification
 *  2022-02-16  Jason Bacon Begin
 ***************************************************************************/

#include <stdio.h>
#include <sysexits.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/param.h>      // PATH_MAX
#include <xtend/mem.h>
#include <xtend/file.h>
#include <xtend/string.h>   // strlcat() on Linux
#include <biolibc/gff.h>
#include "gff-region.h"

int     intersect(int argc, char *argv[]);
void    print_region_feature_names(bl_gff_region_t *region);
void    usage(char *argv[]);

int     main(int argc,char *argv[])

{
    if ( argc < 3 )
	usage(argv);

    return intersect(argc, argv);
}


/***************************************************************************
 *  Description:
 *      
 *  History: 
 *  Date        Name        Modification
 *  2022-02-23  Jason Bacon Begin
 ***************************************************************************/

int     intersect(int argc, char *argv[])

{
    int             arg, count, common_count, c, old_count, new_count;
    bl_gff_region_t r1, rn, *intersect, *new_intersect, *div_intersect;
    char            intersect_file[PATH_MAX + 1];
    FILE            *intersect_stream;
    
    bl_gff_region_init(&r1);
    bl_gff_region_init(&rn);

    printf("%-20s %2s %2s %2s  %s\n", "Species", "Ne", "Co",
	    "Ch", "Intersection with all previous genes");
    for (arg = 1; (arg < argc) &&
		  (common_count = bl_gff_region_load(&r1, argv[arg])) == 0;
		  ++arg)
	printf("%-20s %2c %2c %2c\n", BL_GFF_REGION_SPECIES(&r1),
		'-', '-', '-');
    if ( arg == argc )
	return EX_OK;
    printf("%-20s %2zu %2c %2s  ", BL_GFF_REGION_SPECIES(&r1),
	   BL_GFF_REGION_COUNT(&r1), '*',
	   BL_GFF_REGION_CHROM(&r1));
    puts("(NA)");

    for (++arg; (arg < argc) && (strcmp(argv[arg], "--diverged") != 0) &&
		  (common_count = bl_gff_region_load(&rn, argv[arg])) == 0;
		  ++arg)
	printf("%-20s %2c %2c %2c\n", BL_GFF_REGION_SPECIES(&rn),
		'-', '-', '-');
    if ( arg == argc )
	return EX_OK;
    intersect = bl_gff_region_intersect(&r1, &rn);
    printf("%-20s %2zu %2zu %2s ", BL_GFF_REGION_SPECIES(&rn),
	   BL_GFF_REGION_COUNT(&rn), BL_GFF_REGION_COUNT(intersect),
	   BL_GFF_REGION_CHROM(&rn));
    print_region_feature_names(intersect);
    putchar('\n');
    
    snprintf(intersect_file, PATH_MAX + 1, "Intersects/%s-%s-%s",
	     BL_GFF_REGION_GOI(&r1), BL_GFF_REGION_SPECIES(&r1),
	     BL_GFF_REGION_SPECIES(&rn));
    
    for (++arg; (arg < argc) && (strcmp(argv[arg], "--diverged") != 0); ++arg)
    {
	if ( (count = bl_gff_region_load(&rn, argv[arg])) == 0 )
	    printf("%-20s %2c %2c %2c\n", BL_GFF_REGION_SPECIES(&rn),
		    '-', '-', '-');
	else
	{
	    old_count = BL_GFF_REGION_COUNT(intersect);
	    new_intersect = bl_gff_region_intersect(intersect, &rn);
	    new_count = BL_GFF_REGION_COUNT(new_intersect);
	    bl_gff_region_free(intersect);
	    free(intersect);
	    intersect = new_intersect;
	    
	    printf("%-20s %2zu %2zu %2s ", BL_GFF_REGION_SPECIES(&rn),
		    BL_GFF_REGION_COUNT(&rn),
		    BL_GFF_REGION_COUNT(intersect),
		    BL_GFF_REGION_CHROM(&rn));
	    print_region_feature_names(intersect);
	    putchar('\n');
	    
	    if ( new_count > old_count )
	    {
		fprintf(stderr, "Ooops, new intersect count cannot exceed old!\n");
		exit(EX_SOFTWARE);
	    }

	    // Update GFF filename
	    strlcat(intersect_file, "-", PATH_MAX + 1);
	    strlcat(intersect_file, BL_GFF_REGION_SPECIES(&rn), PATH_MAX + 1);
	    strlcat(intersect_file, ".gff3", PATH_MAX + 1);
	}
    }
    
    // Write GFF for intersect
    xt_rmkdir("Intersects", 0777);
    if ( (intersect_stream = fopen(intersect_file, "w")) == NULL )
    {
	fprintf(stderr, "ms-common-to: Could not open %s for write: %s.\n",
		intersect_file, strerror(errno));
	exit(EX_CANTCREAT);
    }
    fprintf(intersect_stream, "##gff-version 3\n");
    fprintf(intersect_stream, "## %s\n", intersect_file);
    fprintf(intersect_stream, "## This file is simply a list of genes shared by multiple species.\n");
    fprintf(intersect_stream, "## There is no chrom, start, end, etc. since they will differ across species.\n");
    for (c = 0; c < BL_GFF_REGION_COUNT(intersect); ++c)
	bl_gff_write(&BL_GFF_REGION_FEATURES_AE(intersect, c),
		     intersect_stream, BL_GFF_FIELD_ALL);
    fclose(intersect_stream);

    if ( arg < argc )   // Implies --diverged
    {
	++arg;
	puts("\nGenes conserved in each species through divergence from the group above:\n");
	for (; arg < argc; ++arg)
	{
	    if ( (count = bl_gff_region_load(&rn, argv[arg])) == 0 )
		printf("%-20s %2c %2c %2c\n", BL_GFF_REGION_SPECIES(&rn),
			'-', '-', '-');
	    else
	    {
		old_count = BL_GFF_REGION_COUNT(intersect);
		div_intersect = bl_gff_region_intersect(intersect, &rn);
		new_count = BL_GFF_REGION_COUNT(div_intersect);
		
		printf("%-20s %2zu %2zu %2s ", BL_GFF_REGION_SPECIES(&rn),
			BL_GFF_REGION_COUNT(&rn),
			BL_GFF_REGION_COUNT(div_intersect),
			BL_GFF_REGION_CHROM(&rn));
		print_region_feature_names(div_intersect);
		putchar('\n');
		
		if ( new_count > old_count )
		{
		    fprintf(stderr, "Ooops, new intersect count cannot exceed old!\n");
		    exit(EX_SOFTWARE);
		}
	    }
	}
    }

    return EX_OK;
}


/***************************************************************************
 *  Use auto-c2man to generate a man page from this comment
 *
 *  Library:
 *      #include <>
 *      -l
 *
 *  Description:
 *  
 *  Arguments:
 *
 *  Returns:
 *
 *  Examples:
 *
 *  Files:
 *
 *  Environment
 *
 *  See also:
 *
 *  History: 
 *  Date        Name        Modification
 *  2022-03-01  Jason Bacon Begin
 ***************************************************************************/

void    print_region_feature_names(bl_gff_region_t *region)

{
    int         c;
    bl_gff_t    *gff;
    
    for (c = 0; c < BL_GFF_REGION_COUNT(region); ++c)
    {
	gff = &(BL_GFF_REGION_FEATURES_AE(region, c));
	printf(" %s", BL_GFF_FEATURE_NAME(gff));
    }
}


void    usage(char *argv[])

{
    fprintf(stderr, "Usage: %s region.gff3 region.gff3 [...]\n"
		    "       [--diverged region.gff3 [region.gff3 ...]]\n",
	    argv[0]);
    exit(EX_USAGE);
}
